Optimize MVM_string_join in both case of all strands and case of a single element to join

MasterDuke17 opened the pull request:
When everything to be joined are strands of the same type, just copy
them directly instead of creating a single flat string. For small
strings the overhead of being strands counteracts the saving in copying,
but for large strings it's much faster.

For example:
```
my $b = "bbbbbbbb" x 200;
my $c = "cccccccccccccc" x 300;
for ^100_000 {
    $ = join(",", $b, $c, "dddd")
}
```
took ~3.7s before this change and ~1.5s after.

NQP builds ok and passes `make m-test` and Rakudo builds ok and passes `make m-test m-spectest`.

MasterDuke17 commented:
40da3e6 removes the need for https://github.com/rakudo/rakudo/pull/1171 (and still passes all tests and spectests).

jnthn commented on src/strings/ops.c:
```diff
@@ -1630,16 +1631,35 @@ MVMString * MVM_string_join(MVMThreadContext *tc, MVMString *separator, MVMObjec
+                return MVM_repr_get_str(tc, item);
+        }
+    }
+
```
This part of the PR looks just fine.

jnthn commented on src/strings/ops.c:
```diff
@@ -1630,16 +1631,35 @@ MVMString * MVM_string_join(MVMThreadContext *tc, MVMString *separator, MVMObjec
+    /* If the separator is a strand, find out its storage_type. */
+    all_strands_of_same_type = separator->body.storage_type == MVM_STRING_STRAND;
+    if (all_strands_of_same_type)
+        strand_type = separator->body.storage.strands->blob_string->body.storage_type;
```
Why does it matter that all strands have the same storage type? A strand string can reference one 8-bit and another 32-bit wide strand just fine.

MasterDuke17 replied:
Huh, I think I misinterpreted something @timo said (https://irclog.perlgeek.de/perl6-dev/2017-09-23#i_15206788) about doing that as an optimization. I thought he meant make support for mixed strings, but now I see he meant actually make strings with mixed strands, since that is actually supported. Cool, I'll add a commit removing that restriction.

jnthn commented on src/strings/ops.c:
```diff
@@ -1693,13 +1718,27 @@ MVMString * MVM_string_join(MVMThreadContext *tc, MVMString *separator, MVMObjec
+        result->body.num_strands     = total_strands;
+        for (i = 0; i < num_pieces; i++) {
+            MVMString *piece = pieces[i];
+            if (0 < i) {
```
This is one of the places where I really don't find the "always use `<`" thing helps at all...

MasterDuke17 replied:
Personally I agree, but copied the code from what's now the "else" branch and wanted it to be consistent. Should I change it?

jnthn commented on src/strings/ops.c:
```diff
@@ -1693,13 +1718,27 @@ MVMString * MVM_string_join(MVMThreadContext *tc, MVMString *separator, MVMObjec
+                offset += separator->body.num_strands;
+            }
+            copy_strands(tc, piece, 0, result, offset, piece->body.num_strands);
+            offset += piece->body.num_strands;
```
This code doesn't check if the result is NFG stable or not, which means it'll produce wrong results if two of the things being joined would form a single grapheme.

MasterDuke17 replied:
I wondered about this, but couldn't craft an example where the behavior after my change did anything different than the code before. I tried various combinations of joining elements that had strands of just combining characters, just combining characters as the separator, etc. Also, it does pass a spectest. Either the collapsing is correctly happening elsewhere, my ad-hoc test cases weren't actually testing the right things, or we should add some tests for this to roast. Do you have an example I can test?

MasterDuke17 commented:
@jnthn removed the bit about all strands having to have the same storage_type and added in the concat stability checking code from the not-all-strands else branch.

With @samcv++'s help, added in some debugging code that did get tripped up with the initial changes, but now is clean.

samcv commented on src/strings/ops.c:
```diff
@@ -1693,13 +1714,51 @@ MVMString * MVM_string_join(MVMThreadContext *tc, MVMString *separator, MVMObjec
-    if (total_strands < MVM_STRING_MAX_STRANDS && 16 <= total_graphs / total_strands) {
-        /* XXX TODO: Implement this, conditionalize branch thing below. */
+    /* If the separator and pieces are all strands, and there are
+     * enough strands of a large enough size, just copy them directly. */
```
Can you change it to this please:
```
/* If the separator and pieces are all strands, and
 * there are on average at least 16 graphemes in each of the strands */
if (all_strands && total_strands < MVM_STRIND_MAX_STRANDS
           && total_strands * 16 <= total_graphs)
```
The conditional becomes much clearer and clearer comment.

MasterDuke17 commented:
@samcv and done.

samcv commented:
Thank you :-)
